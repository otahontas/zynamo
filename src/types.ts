/* eslint-disable @typescript-eslint/no-explicit-any */
type Assert<T extends true> = T;
type Equals<A, B> = A extends B ? (B extends A ? true : false) : false;

// Typed Errors
type ParseErrorTag = "Parsing Error:";
type CreateParseError<Content extends string> = `${ParseErrorTag} ${Content}`;

type KeyConditionExpressSyntaxError<Expression extends string> =
  `${Expression} was not proper KeyConditionExpression`;

type ValidatePartitionKeyPart<
  PartitionKeyPart extends string,
  PartitionKey extends string,
  ExpressionAttributevalues extends Record<string, any>
> = PartitionKeyPart extends `${infer PartitionKeyName} = ${infer PartitionKeyVariableName}`
  ? // Check if partition key name is correct
    PartitionKeyName extends PartitionKey
    ? // Check if variable is name is in ExpressionAttributevalues
      PartitionKeyVariableName extends keyof ExpressionAttributevalues
      ? PartitionKeyPart
      : CreateParseError<`Partition key variable ${PartitionKeyVariableName} was not in ExpressionAttributevalues`>
    : CreateParseError<`Partition key in expression should be ${PartitionKey}, but it was ${PartitionKeyName}`>
  : CreateParseError<`Partition key part '${PartitionKeyPart}' of expression was not properly formatted`>;

type SortKeySeparator = "=" | "<" | "<=" | ">" | ">=";

type ValidateSortKeyPart<
  SortKeyPart extends string,
  SortKey extends string,
  ExpressionAttributevalues extends Record<string, any>
> =
  // Check if we have a sort key and one variable
  SortKeyPart extends
    | `${infer SortKeyName} ${SortKeySeparator} ${infer SortKeyVariableName}`
    | `begins_with (${infer SortKeyName}, ${infer SortKeyVariableName})`
    ? // Check if sort key name is correct
      SortKeyName extends SortKey
      ? // Check if sort key variable name is in ExpressionAttributevalues
        SortKeyVariableName extends keyof ExpressionAttributevalues
        ? SortKeyPart
        : CreateParseError<`Sort key variable ${SortKeyVariableName} was not in ExpressionAttributevalues`>
      : CreateParseError<`Sort key in expression should be ${SortKey}, but it was ${SortKeyName}`>
    : // Check if we have a sort key and two variables
    SortKeyPart extends `${infer SortKeyName} BETWEEN ${infer SortKeyVariableName1} AND ${infer SortKeyVariableName2}`
    ? // Check if sort key name is correct
      SortKeyName extends SortKey
      ? // Check if both sort variables names are in ExpressionAttributevalues
        SortKeyVariableName1 extends keyof ExpressionAttributevalues
        ? SortKeyVariableName2 extends keyof ExpressionAttributevalues
          ? SortKeyPart
          : CreateParseError<`First sort key variable ${SortKeyVariableName1} was not in ExpressionAttributevalues`>
        : CreateParseError<`Second sort key variable ${SortKeyVariableName2} was not in ExpressionAttributevalues`>
      : CreateParseError<`Sort key in expression should be ${SortKey}, but it was ${SortKeyName}`>
    : CreateParseError<`Sort key part '${SortKeyPart}' of expression was not properly formatted`>;

export type ValidateKeyConditionExpression<
  Expression extends string,
  PartitionKey extends string,
  ExpressionAttributevalues extends Record<string, any>,
  SortKey extends string | undefined = undefined
> =
  // Check if we have have both partition key and sort key
  Expression extends `${infer PartitionKeyPart} AND ${infer SortKeyPart}`
    ? SortKey extends string
      ? // Check if both partition key and sort key are correct
        ValidatePartitionKeyPart<
          PartitionKeyPart,
          PartitionKey,
          ExpressionAttributevalues
        > extends infer PartitionKeyValidationResult
        ? PartitionKeyValidationResult extends `${ParseErrorTag}${string}`
          ? // Not valid partition key, return error
            PartitionKeyValidationResult
          : ValidateSortKeyPart<
              SortKeyPart,
              SortKey,
              ExpressionAttributevalues
            > extends infer SortKeyValidationResult
          ? SortKeyValidationResult extends `${ParseErrorTag}${string}`
            ? // Not valid sort key, return error
              SortKeyValidationResult
            : // All okay, return expression
              Expression
          : never
        : never
      : "Sort key parameter must be defined if expression contains sort key"
    : // Validate if expression is the partitionkey part
    Equals<
        ValidatePartitionKeyPart<
          Expression,
          PartitionKey,
          ExpressionAttributevalues
        >,
        Expression
      > extends true
    ? // Return true expression if valid
      Expression
    : // Return error if invalid
      KeyConditionExpressSyntaxError<Expression>;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type _goodCases = [
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode = :e",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode = :e"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode < :e",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode < :e"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode <= :e",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode <= :e"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode > :e",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode > :e"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode >= :e",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode >= :e"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND Episode BETWEEN :e1 AND :e2",
        "Season",
        { ":e1": 1; ":e2": 2; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND Episode BETWEEN :e1 AND :e2"
    >
  >,
  Assert<
    Equals<
      ValidateKeyConditionExpression<
        "Season = :s AND begins_with (Episode, :e)",
        "Season",
        { ":e": 1; ":s": 1 },
        "Episode"
      >,
      "Season = :s AND begins_with (Episode, :e)"
    >
  >
];
