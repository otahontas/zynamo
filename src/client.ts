import {
  CreateTableCommand,
  CreateTableCommandOutput,
  DeleteTableCommand,
  DeleteTableCommandOutput,
  DynamoDBClient,
  DynamoDBClientConfig,
} from "@aws-sdk/client-dynamodb";
import {
  DynamoDBDocumentClient,
  GetCommand,
  GetCommandOutput,
  PutCommand,
  PutCommandOutput,
  type TranslateConfig,
} from "@aws-sdk/lib-dynamodb";
import type { HttpHandlerOptions } from "@aws-sdk/types";
import type { z } from "zod";
import type { Spread } from "type-fest";

// TS doesn't allow classes to override static methods of base classes: hence
// omitting "from" method her
// https://github.com/microsoft/TypeScript/issues/4628#issuecomment-1147905253
type DynamoDBDocumentClientConstructorParameters = [
  client: DynamoDBClient,
  translateConfig?: TranslateConfig
];
// @ts-expect-error DynamoDBDocumentClient constructor is protected, but needs to used
// here to omit "from"
const DynamoDBDocumentClientWithoutFrom: (new (
  ...arguments_: DynamoDBDocumentClientConstructorParameters
) => DynamoDBDocumentClient) &
  Omit<typeof DynamoDBDocumentClient, "from"> = DynamoDBDocumentClient;

// TODO: Figure out a way to overload only the modified stuff, no other fields

type ZynamoCommand =
  | GetCommand
  | PutCommand
  | CreateTableCommand
  | DeleteTableCommand;
type ZynamoOutputs =
  | GetCommandOutput
  | PutCommandOutput
  | CreateTableCommandOutput
  | DeleteTableCommandOutput;

// TODO: what
// type ValidPutCommandInput<Schema extends z.AnyZodObject> = Spread<
//   Omit<PutCommandInput, "Item">,
//   { Item: z.input<Schema> }
// >;

// const ValidPutCommand: (new (
//   ...args: ConstructorParameters<typeof PutCommand>
// ) => PutCommand) &
//   Omit<typeof PutCommand, "input"> & { input: ValidPutCommandInput } =
//   PutCommand;

export class ZynamoClient<
  Schema extends z.AnyZodObject
> extends DynamoDBDocumentClientWithoutFrom {
  schema: Schema;

  protected constructor(
    schema: Schema,
    ...superParameters: DynamoDBDocumentClientConstructorParameters
  ) {
    super(...superParameters);
    this.schema = schema;
  }

  // === Implementations for overridden functionalities
  async handleGetCommand(command: GetCommand) {
    const result = await super.send(command);
    if (result.Item) {
      return {
        ...result,
        // This gets inferred correctly
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        Item: this.schema.parse(result.Item),
      };
    }
    return result;
  }

  async handlePutCOmmand(command: PutCommand) {
    this.schema.parse(command.input.Item);
    return super.send(command);
  }

  // === Overloads for overridden functionalities
  override send(
    command: GetCommand
  ): Promise<
    Spread<Omit<GetCommandOutput, "Item">, { Item?: z.output<Schema> }>
  >;

  override send(command: PutCommand): Promise<PutCommandOutput>;

  // === Not overridden but I want to keep these usable
  override send(command: CreateTableCommand): Promise<CreateTableCommandOutput>;

  override send(command: DeleteTableCommand): Promise<DeleteTableCommandOutput>;

  override send(
    command: ZynamoCommand,
    options?: HttpHandlerOptions
  ): Promise<ZynamoOutputs> {
    if (command instanceof GetCommand) {
      return this.handleGetCommand(command);
    }
    if (command instanceof PutCommand) {
      return this.handlePutCOmmand(command);
    }
    // TODO: FIX
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument
    return super.send(command as any, options) as any;
  }

  // Create dynamoDBClient and pass it to ZynamoClient like with documentClient
  static from<Schema extends z.AnyZodObject>(
    schema: Schema,
    config: DynamoDBClientConfig,
    translateConfig?: TranslateConfig
  ) {
    const dynamoDBClient = new DynamoDBClient(config);
    return new ZynamoClient<Schema>(schema, dynamoDBClient, translateConfig);
  }
}
