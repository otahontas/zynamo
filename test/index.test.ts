import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { GenericContainer, type StartedTestContainer } from "testcontainers";
import {
  CreateTableCommand,
  DeleteTableCommand,
  type DynamoDBClientConfig,
} from "@aws-sdk/client-dynamodb";
import { z } from "zod";
import { GetCommand, PutCommand } from "@aws-sdk/lib-dynamodb";
import { generateMock } from "@anatine/zod-mock";
import { ZynamoClient } from "../src";

const dynamodbPort = 8000;

describe("zynamo", () => {
  let container: StartedTestContainer;
  let containerUrl: string;
  let config: DynamoDBClientConfig;

  beforeAll(async () => {
    console.log("Setting up dynamodb container");
    container = await new GenericContainer("amazon/dynamodb-local")
      .withExposedPorts(dynamodbPort)
      .start();

    const logs = await container.logs();
    logs.on("err", (line) => console.error("[Container error]:", line));

    containerUrl = `http://${container.getHost()}:${container.getMappedPort(
      dynamodbPort
    )}`;

    config = {
      endpoint: containerUrl,
      region: "us-east-1",
      credentials: {
        accessKeyId: "xxxxx",
        secretAccessKey: "xxxxx",
      },
    };

    console.log(
      `Dynamodb container with id ${container.getId()} and name ${container.getName()} running on ${containerUrl}}`
    );
  });

  afterAll(async () => {
    await container.stop();
  });

  describe("with schema", () => {
    // TODO: better tests with some property based hassle (all different data types for
    // dynamo etc)
    const schema = z.object({
      foo: z.string(),
      bar: z.number(),
      foobar: z.boolean(),
      foobarbaz: z.string().optional(),
      foobarbazqux: z.string().nullable(),
      foobaaars: z.array(z.string()),
    });
    let client: ZynamoClient<typeof schema>;
    beforeAll(() => {
      client = ZynamoClient.from(schema, config);
    });

    it("client can be created with proper schema", () => {
      const clientHere = ZynamoClient.from(schema, config);
      expect(clientHere).toBeDefined();
      // TODO: test that client doesn't accept unsuitable schema
    });

    it("client can create a table", async () => {
      // TODO: test that client accepts only table names defined in schema (later when
      // schemas handle multiple tables & sort keys etc)
      //
      const tableName = "FooTable";
      const response = await client.send(
        new CreateTableCommand({
          AttributeDefinitions: [
            {
              AttributeName: "Swag",
              AttributeType: "S",
            },
          ],
          KeySchema: [
            {
              AttributeName: "Swag",
              KeyType: "HASH",
            },
          ],
          ProvisionedThroughput: {
            ReadCapacityUnits: 1,
            WriteCapacityUnits: 1,
          },
          TableName: "FooTable",
        })
      );
      expect(response.$metadata.httpStatusCode).toBe(200);
      await client.send(new DeleteTableCommand({ TableName: tableName }));
    });

    describe("putting data to table", () => {
      const tableForPut = "FooTableToPut";
      beforeAll(async () => {
        await client.send(
          new CreateTableCommand({
            AttributeDefinitions: [
              {
                AttributeName: "foo",
                AttributeType: "S",
              },
            ],
            KeySchema: [
              {
                AttributeName: "foo",
                KeyType: "HASH",
              },
            ],
            ProvisionedThroughput: {
              ReadCapacityUnits: 1,
              WriteCapacityUnits: 1,
            },
            TableName: tableForPut,
          })
        );
      });

      afterAll(async () => {
        await client.send(new DeleteTableCommand({ TableName: tableForPut }));
      });

      it("succeeds with valid input", async () => {
        const input = generateMock(schema);
        const response = await client.send(
          new PutCommand({
            TableName: tableForPut,
            Item: input,
          })
        );
        expect(response.$metadata.httpStatusCode).toBe(200);
      });
      it("fails with not valid input", () => {
        void expect(
          client.send(
            new PutCommand({
              TableName: tableForPut,
              Item: {
                swag: "littinen",
              },
            })
          )
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });

    describe("getting data from table", () => {
      const tableForGet = "FooTableForGet";
      const fooValue = "1";
      let input: z.infer<typeof schema>;
      beforeAll(async () => {
        await client.send(
          new CreateTableCommand({
            AttributeDefinitions: [
              {
                AttributeName: "foo",
                AttributeType: "S",
              },
            ],
            KeySchema: [
              {
                AttributeName: "foo",
                KeyType: "HASH",
              },
            ],
            ProvisionedThroughput: {
              ReadCapacityUnits: 1,
              WriteCapacityUnits: 1,
            },
            TableName: tableForGet,
          })
        );
        // add input
        input = generateMock(schema);
        input.foo = fooValue;
        await client.send(
          new PutCommand({
            TableName: tableForGet,
            Item: input,
          })
        );
      });

      afterAll(async () => {
        await client.send(new DeleteTableCommand({ TableName: tableForGet }));
      });

      it("returns valid stuff", async () => {
        const response = await client.send(
          new GetCommand({
            TableName: tableForGet,
            Key: {
              foo: fooValue,
            },
          })
        );
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(response.Item).toMatchObject(input);
      });

      it("doesn't fail when item not in table", async () => {
        const response = await client.send(
          new GetCommand({
            TableName: tableForGet,
            // TODO: validate key
            Key: {
              foo: "nonValidValue",
            },
          })
        );
        expect(response.$metadata.httpStatusCode).toBe(200);
        expect(response.Item).toBeUndefined();
      });
      // TODO: test that we error for data integrity when data doesn't match schema
      // content
    });
  });
});
