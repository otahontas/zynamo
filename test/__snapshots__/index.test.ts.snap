// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`zynamo > with schema > putting data to table > fails with not valid input 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"foo\\"
    ],
    \\"message\\": \\"Required\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"bar\\"
    ],
    \\"message\\": \\"Required\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"boolean\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"foobar\\"
    ],
    \\"message\\": \\"Required\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"foobarbazqux\\"
    ],
    \\"message\\": \\"Required\\"
  },
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"array\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"foobaaars\\"
    ],
    \\"message\\": \\"Required\\"
  }
]"
`;
